src/compiler/python_generator.cc

```diff
@@ -173,11 +116,62 @@ grpc::string ModuleAlias(const grpc::string& filename) {
   return module_name;
 }
 
-bool GetModuleAndMessagePath(const Descriptor* type,
-                             const ServiceDescriptor* service,
-                             grpc::string* out) {
+// Tucks all generator state in an anonymous namespace away from
+// PythonGrpcGenerator and the header file, mostly to encourage future changes
+// to not require updates to the grpcio-tools C++ code part. Assumes that it is
+// only ever used from a single thread.
+struct PrivateGenerator {
+  const GeneratorConfiguration& config;
+  const FileDescriptor *file;
+
+  bool split_pb2_grpc;
+
+  Printer *out;
+
+  PrivateGenerator(const GeneratorConfiguration& config,
+                   const FileDescriptor *file);
+
+  std::pair<bool, grpc::string> GetGrpcServices();
+
+ private:
+  bool PrintPreamble();
+  bool PrintPreambledGAServices();
```

Comment:
I think this could be more clear if this was just PrintServices(), as opposed to bundling preamble logic with the services.